RInAppMessaging/Classes/Router.swift

```diff
@@ -14,27 +14,36 @@ internal protocol RouterType: AnyObject {
                          completion: @escaping (_ cancelled: Bool) -> Void)
 
     /// Removes displayed campaign view from the stack
-    func discardDisplayedCampaign()
+    /// - Returns: a campaign associated with displayed message or nil if no message was displayed
+    func discardDisplayedCampaign() -> Campaign?
 }
 
 /// Handles all the displaying logic of the SDK.
 internal class Router: RouterType {
 
     private let dependencyManager: DependencyManager
+    private let displayQueue = DispatchQueue(label: "IAM.MessageLoader")
     var accessibilityCompatibleDisplay = false
 
     init(dependencyManager: DependencyManager) {
         self.dependencyManager = dependencyManager
     }
 
-    func discardDisplayedCampaign() {
-        DispatchQueue.main.async {
-            guard let rootView = UIApplication.shared.getKeyWindow() else {
-                return
+    func discardDisplayedCampaign() -> Campaign? {
+        return displayQueue.sync {
+            let discard: () -> Campaign? = {
+                guard let rootView = UIApplication.shared.getKeyWindow() else {
+                    return nil
+                }
+
+                let presentedView = rootView.findIAMViewSubview()
+                presentedView?.onDismiss?(true)
+                presentedView?.removeFromSuperview()
+
+                return presentedView?.basePresenter.campaign
             }
 
-            let presentedView = rootView.findIAMViewSubview()
-            presentedView?.removeFromSuperview()
+            return Thread.isMainThread ? discard() : DispatchQueue.main.sync(execute: discard)
```

Comment:
If you're running on displayQueue can the `Thread.isMainThread` check ever evaluate true?

Reply:
This is mainly for unit tests

Reply:
Ah, I know what you mean now. It's because it's `displayQueue.sync()` not `async()`. Without that check unit tests that executed `discardDisplayedCampaign()` on main thread got deadlocked. Normally this method is called on main IAM queue.